crates/matrix-sdk/src/send_queue/upload.rs

```diff
@@ -398,4 +402,132 @@ impl QueueStorage {
 
         Ok(())
     }
+
+    /// Try to abort an upload that would be ongoing.
+    ///
+    /// Return true if any media (media itself or its thumbnail) was being
+    /// uploaded. In this case, the media event has also been removed from
+    /// the send queue. If it returns false, then the uploads already
+    /// happened, and the event sending *may* have started.
+    #[instrument(skip(self, handles))]
+    pub(super) async fn abort_upload(
+        &self,
+        event_txn: &TransactionId,
+        handles: &MediaHandles,
+    ) -> Result<bool, RoomSendQueueStorageError> {
+        let client = self.client()?;
+
+        // Keep the lock until we're done touching the storage.
+        let mut being_sent = self.being_sent.write().await;
+        debug!("trying to abort an upload");
+
+        let store = client.store();
+
+        let upload_file_as_dependent = ChildTransactionId::from(handles.upload_file_txn.clone());
+        let event_as_dependent = ChildTransactionId::from(event_txn.to_owned());
+
+        let mut removed_dependent_upload = false;
+        let mut removed_dependent_event = false;
+
+        if let Some(thumbnail_txn) = &handles.upload_thumbnail_txn {
+            if store.remove_send_queue_request(&self.room_id, &thumbnail_txn).await? {
+                // The thumbnail upload existed as a request: either it was pending (something
+                // else was being sent), or it was actively being sent.
+                trace!("could remove thumbnail request, removing 2 dependent requests now");
+
+                // 1. Try to abort sending using the being_sent info, in case it was active.
+                if let Some(info) = being_sent.as_ref() {
+                    if info.transaction_id == *thumbnail_txn {
+                        // SAFETY: we knew it was Some(), two lines above.
+                        let info = being_sent.take().unwrap();
```

Comment:
Rookie Rust dev question: why do we need another `info` here?

Reply:
Good question! Above it was taken `as_ref()`, so a `&BeingSentInfo`; the cancellation requires triggering the `oneshot::channel`, which consumes it by ownership, so we need to take `BeingSentInfo` by ownership here; at this point it's fine because we know it's the last use. I'll add a comment and maybe a tiny function helper to make this clearer.

Reply:
Ah, I see, I didn't realise the channel is consumed 👌 .